File: Smx.Vst/Smx/Filter.cs
using Jacobi.Vst.Plugin.Framework;
using Smx.Vst.Util;
using System;

namespace Smx.Vst.Smx
{
  internal class Filter
  {
    public class FilterParameter
    {
      public FilterParameter(VstParameterManager modeMgrs,
                             VstParameterManager cutoffMgr,
                             VstParameterManager dryWetMgr)
      {
        ModeMgrs = modeMgrs;
        CutoffMgr = cutoffMgr;
        DryWetMgr = dryWetMgr;
      }

      public VstParameterManager CutoffMgr { get; private set; }
      public VstParameterManager DryWetMgr { get; private set; }
      public VstParameterManager ModeMgrs { get; private set; }
    }

    private double buf0;

    private double buf1;
    private readonly FilterParameter parameters;

    //private double cutoff;

    //private double feedbackMount;

    //private double dryWet;

    //private Mode mode;

    public Filter(FilterParameter parameters)
    {
      this.parameters = parameters;
    }

    public enum Mode
    {
      None,
      Lowpass,
      Highpass,
      Bandpass
    }

    public double Process(double inputValue)
    {
      return AudioMath.Mix(inputValue, ProcessInternal(inputValue), parameters.DryWetMgr.CurrentValue);
    }

    private double ProcessInternal(double inputValue)
    {
      var cutoff = Math.Pow(parameters.CutoffMgr.CurrentValue, 4.0);
      //buf0 += cutoff * (inputValue - buf0 + feedbackAmount * (buf0 - buf1));
      buf0 += cutoff * (inputValue - buf0);
      buf1 += cutoff * (buf0 - buf1);
      switch ((Mode)parameters.ModeMgrs.CurrentValue)
      {
        case Mode.Lowpass:
          return buf1;

        case Mode.Highpass:
          return inputValue - buf0;

        case Mode.Bandpass:
          return buf0 - buf1;

        default:
          return 0.0;
      }
    }
  }
}

File: Smx.Native.Managed/Filter.cpp
#include "Filter.h"
#include "AudioMath.h"

Filter::Filter(FilterParameter^ parameters) {
  this->parameters = parameters;
}

double Filter::Process(double inputValue) {
  FilterMode mode = static_cast<FilterMode>(parameters->Mode);
  double wet = ProcessInternal(inputValue, mode);
  double wetAmt = parameters->WetAmt;
  return (mode >= FilterMode::LowpassAdd) ? (inputValue + wet * wetAmt)
    : (mode >= FilterMode::LowpassMix) ? AudioMath::Mix(inputValue, wet, wetAmt)
    : inputValue;
}


double Filter::CalcFeedbackAmount(double cutoff) {
  double feedback = static_cast<double>(parameters->Resonance);
  feedback += feedback / (1.0 - AudioMath::Clamp(cutoff, 0.99, 0.01));
  return feedback;
}

double Filter::ProcessInternal(double inputValue, FilterMode mode) {
  double cutoff = pow(parameters->Cutoff, 4.0);
  double feedback = CalcFeedbackAmount(cutoff);

  buf0 += cutoff * (inputValue - buf0 + feedback * (buf0 - buf1));
  buf1 += cutoff * (buf0 - buf1);

  switch (mode) {
  case FilterMode::LowpassMix:
  case FilterMode::LowpassAdd:
    return buf1;

  case FilterMode::HighpassMix:
  case FilterMode::HighpassAdd:
    return inputValue - buf0;

  case FilterMode::BandpassMix:
  case FilterMode::BandpassAdd:
    return buf0 - buf1;

  default:
    return inputValue;
  }
}